### PR TITLE
XRT-310 Save calibration Cache XOCL_CALIB_STORAGE

### PR DESCRIPTION
**Pull request by chienwei-lan**
Save each calibration cache to XOCL_CALIB_STORAGE sub-device.

1. Try to restore calibration cache back to SRSR IP before we kick start calibration even it's empty at first time. 

If the fast calibration fails(calibration cache is empty), it will fall back to full calibration anyway.

2. Save the calibration cache after calibration complete.

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 1]
```diff
@@ -1956,9 +1956,23 @@ static int icap_create_post_download_subdevs(struct platform_device *pdev, struc
...
+			left_parentness = strstr(ip->m_name, "BANK");
+			right_parentness = left_parentness+5;
+			digit_len = 1;
+			strncpy(temp, left_parentness+4, digit_len);
+			temp[digit_len] = '\0';
+
+			if (kstrtoint(temp, 10, &idx) != 0) {
+				err = -EINVAL;
+				goto done;
+			}
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: How about simplify the code by sscanf?

Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: done!

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c [hunk 2]
```diff
@@ -78,15 +76,14 @@ static int srsr_full_calibration(struct platform_device *pdev)
 	xocl_dr_reg_write32(xdev, CTRL_BIT_SYS_RST, xocl_ddr_srsr->base+REG_CTRL_OFFSET);
 	xocl_dr_reg_write32(xdev, 0x0, xocl_ddr_srsr->base+REG_CTRL_OFFSET);
 
-	for (;i<100;++i) {
+	for (; i < 100; ++i) {
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: please define a macro for constant value like 100 for code readability.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c [hunk 3]
```diff
@@ -126,7 +112,7 @@ static int srsr_save_calib(struct platform_device *pdev)
 	}
 	xocl_dr_reg_write32(xdev, CTRL_BIT_SREF_REQ | CTRL_BIT_XSDB_SELECT, xocl_ddr_srsr->base+REG_CTRL_OFFSET);
 
-	for ( i = 0; i < 4; ++i) {
+	for (i = 0; i < 4; ++i) {
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: need a macro for 4. what is 4 here?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c [hunk 4]
```diff
@@ -154,7 +140,7 @@ static int srsr_fast_calib(struct platform_device *pdev, bool retention)
 		write_val |= CTRL_BIT_MEM_INIT_SKIP;
 
 	xocl_dr_reg_write32(xdev, write_val, xocl_ddr_srsr->base+REG_CTRL_OFFSET);
-	for ( i = 0; i < 4; ++i) {
+	for (i = 0; i < 4; ++i) {
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: need a macro for 4

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/calib_storage.c [hunk 5]
```diff
@@ -50,6 +51,9 @@ static int calib_storage_save_by_idx(struct platform_device *pdev, uint32_t idx)
 		goto done;
 	}
 
+	err = xocl_srsr_read_calib(xdev, idx, calib_storage->cache[idx]->data,
+					XOCL_CALIB_CACHE_SIZE);
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/calib_storage.c]: why do you have a constant value for XOCL_CALIB_CACHE_SIZE? Shouldn't it be the number of memory banks?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c (hunk 2 above)
Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: done, replace with FULL_CALIB_TIMEOUT 100,

full calibration should finish in 2000ms.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c (hunk 3 above)
Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: The calibration cache size is up to 2901 words(so far), means ~12KB, and XSDB size is 4KB.
So we have to read the calibration data from XSDB at least 3 times.

Let's make it to 4, then we have 16kB calibration cache.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/calib_storage.c (hunk 5 above)
Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/calib_storage.c]: calib_storage is the super set of all the calibration cache and make it to an array of struct calib_cache *, each struct calib_cache save 1. mem idx, 2. the 16KB calibration data.

We may need additional field: mem_type to indicate the calibration data belongs to,
just like what we have in mig. 

We can set up a quick call for review the code!

Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/calib_storage.c]: done! get cache_size by asking SRSR_IP

**Pull request merged**